### PR TITLE
Add functional tests for start command

### PR DESCRIPTION
**Pull request by devimc**
Add some functional tests for start command

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+  $ RUNTIME="cc-runtime" make functional
+
+in the above example the installed version of the Runtime is used
+
+
+Environment variables
+---------------------
+
+- `RUNTIME` - Path of Clear Containers Runtime, the default path is ``$GOPATH/src/github.com/clearcontainers/runtime/cc-runtime``
```
Comment (dlespiau) [inline README.rst]: That documentation seems out of date.

File: functional/start_test.go [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+package functional
+
+import (
+	. "github.com/clearcontainers/tests/lib"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("start command", func() {
```
Comment (dlespiau) [inline functional/start_test.go]: Maybe just "start"?

File: functional/start_test.go [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+	c := NewCommand(Runtime, "start")
+	c.ExitCode = 1
+	ret := c.Run()
+	Context("without container id", func() {
+		It("should NOT return 0", func() {
+			Expect(ret).NotTo(Equal(0))
+		})
+		It("should report an error", func() {
```
Comment (dlespiau) [inline functional/start_test.go]: probably a good idea to match certain key words in the error message

File: lib/utils.go [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+	// ExitCode is the expected exit code
+	ExitCode int
+}
+
+func init() {
+	flag.StringVar(&Runtime, "runtime", "cc-runtime", "Path of Clear Containers Runtime")
+	flag.StringVar(&Proxy, "proxy", "cc-proxy", "Path of Clear Containers Proxy")
+	flag.StringVar(&Shim, "shim", "cc-shim", "Path of Clear Containers Shim")
+	flag.IntVar(&Timeout, "timeout", 5, "Time limit for each test")
```
Comment (dlespiau) [inline lib/utils.go]: Better indicate the unit of this 5

File: lib/utils.go [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+// Runtime path of Clear Containers Runtime
+var Runtime string
+
+// Proxy path of Clear Containers Proxy
+var Proxy string
+
+// Shim path of Clear Containers Shim
+var Shim string
+
+// Timeout path of Clear Containers Shim
```
Comment (dlespiau) [inline lib/utils.go]: TImeout duration (in s)

File: lib/utils.go [hunk 6]
```diff
@@ -0,0 +1,108 @@
+// Copyright (c) 2017 Intel Corporation
```
Comment (dlespiau) [inline lib/utils.go]: Can we find something more descriptive than lib and utils? could be run.go

File: README.rst (hunk 1 above)
Comment (dlespiau) [inline README.rst]: Ah no sorry, the default values are from the Makefile, not the default values of the arguments. Confusing.

**Comment by dlespiau** [comment]
- I'm not too sure what functional tests with shim and proxy would mean. Any idea?
- Functional tests should probably be with each component, no?
- integration tests could be in a separate repo, they bind all components together.
- I think integration tests should be done on already installed and setup components (say package each component, install them, run docker tests, run cri-o tests, run k8s tests, ...).

**Comment by dlespiau** [comment]
Probably a good argument for having functional tests with the components is to be able to change both code and test in 1 commit. Otherwise, if you change the runtime to behave differently, I don't see any practical way we can make the tests pass if they are in a different repo.

File: functional/start_test.go (hunk 3 above)
Comment (devimc) [inline functional/start_test.go]: these test should be compatible with CC 2.0, not sure if we can match certain key words

File: lib/utils.go (hunk 4 above)
Comment (devimc) [inline lib/utils.go]: yes

File: lib/utils.go (hunk 5 above)
Comment (devimc) [inline lib/utils.go]: fixing

**Comment by devimc** [comment]
the idea is to have only one repo for all test, in that way we could share tests between projects
for me functional tests are those tests used to check the functionality of a project (in our case Clear Container) this includes all components working together and integration tests for me are those tests used to check the integration of a project with other projects, projects that "in theory" we can touch/modify easily
I'd like to see the point of view of QA @chavafg

**Comment by sboeuf** [comment]
I agree that I don't get why we are separating the source code from its integration tests. We have 4 repos (virtcontainers, cc-shim, cc-proxy and cc-runtime), and we already have basic unit testing for all those components, and there is room for improvement in all of them, but we want to stay in those repos for adding the tests.
Then, we have to ask ourselves about what we want to add as integration tests for each repo:
- cc-shim: Archana already implemented a mock proxy so that we can test the shim without the entire chain. I don't think there is more to do as this would mean using the real proxy, which is more appropriate for runtime testing.
- cc-proxy: Tests relies on a mock hyperstart to help us testing this component. I think the mock hyperstart could be definitely improved to be more accurate about the behavior we are trying to mock. And on the other side, the shim interaction is reproduced with the client library provided by Damien. But as for the shim, no need to test with entire chain because this is the purpose of the runtime tests.
- virtcontainers: Lots of unit tests but there is room for more. Very basic (we have to add a lot more) benchmarks running real VMs, meaning they have to run in Semaphore CI. Missing complete functional testing, it could be done by creating some scripts using virtc CLI, and/or it could be done by adding unit tests interacting almost as a runtime implementation (but this case means we cannot use travis anymore, it won't support it).
- cc-runtime: We have a few unit tests but the coverage MUST be increased. We have also recently included Semaphore CI support so that we can setup the entire chain and test a new PR is not breaking anything. There is currently only one test
```
$ sudo docker run busybox ifconfig
```
and we have to increase those tests. This is exactly where we want those integration/functional tests to be. Only need to call a script calling all the tests you want from Semaphore CI. Let's try to keep the amount of commands very simple in Semaphore CI.

@devimc @dlespiau @sameo @amshinde let me know if you think something is wrong about this explanation.

**Comment by devimc** [comment]
functional and integration tests involve the use of all components (virtcontainers, cc-shim, cc-proxy and cc-runtime) why should these tests only be part of only one repo? 
we could easily integrate these test with the CI of each repo to ensure its "functionality" and "integration" of that components with other components

**Comment by chavafg** [comment]
The idea of having a separate repository for these functional and integration tests is that they verify that all the components (cc-shim, cc-proxy, virtcontainers and runtime) work correctly and every change on each component should be tested against these tests. These are a more high level functional tests (maybe we can call them system level tests as they integrate different components).

In addition to these tests, we may have other unit or functional tests for each of the repositories we have, and their scope should be to verify the correct functionality of the component.

**Comment by sameo** [comment]
I can't see the diff right now, but the idea of this repo is to be able to separate our tests from their code base in order to run both functional and integration tests on each PR for each component. Also they should be flexible enough to be used by 2.1 or any OCI compatible runtime for that matter.

Both our functional and integration tests rely completely or partially on our proxy, shim, and runtime components. So any change to any of those should trigger those tests and that's why they have to live in a separate repo (both functional and integration ones).

Yes, the drawback is that we won't be able to push component and test changes at the same time but I suspect this will be rarely problematic.

**Comment by dlespiau** [comment]
It makes sense for tests that depend on having everything (but then I'd argue we need to run those tests on the installed version of the components, not on the uninstalled versions like presented here).

Maybe the confusion here was that the example test included in the PR was a runtime test that didn't need shim nor proxy.

**Comment by dlespiau** [comment]
re the dependency between tests and code tested, the failing case is:

1. we have a test in this repo, the test is passing: for instance we test the runtime output a certain string:
2. we send a PR for the runtime that changes that string
3. the runtime PR will fail the tests

I expect that case to be somewhat common, but I guess there's no real better option. It could be possible to indicate a branch of the tests repository to use when sending a PR to runtime/proxy/shim is that's really a problem.

**Comment by sboeuf** [comment]
I would summarize that we will probably end up with cases more complicated than if we had those test in sources repositories, but if we consider that sharing the tests between both versions of the project is a high priority, there is nothing else we can do.

**Comment by sameo** [comment]
Yes, it will introduce some complexity but maintaining duplicates of the same tests is not a better solution I think.

**Comment by sboeuf** [comment]
Yep agreed, there is no perfect solution unfortunately... What about moving those tests when 3.0 will replace 2.1 ? This github.com/clearcontainers/tests could be only temporary.

**Comment by dlespiau** [comment]
Let's talk about it tomorrow, but just wanted to leave a last comment here: it maybe worse than my first example. Imagine wanting to merge a new runtime feature with the corresponding tests in this repo.

- we have two PRs, one in runtime/ and one in tests/
- we can't merge the one in runtime, we don't know if the corresponding tests pass in semaphore. that's a real risk as semaphore is a different environment than the one the developer has tested its code with.
- we can't really merge the tests/ PR for the same reasons. The tests/ PR will also fail because the code needed isn't merged

I wouldn't like to go back to a situation where we have to merge code not tested in CI.

All this can be solved though. I think we need to have the option to specify specific branches of the projects we want to use to test a PR. That way, the testing framework can checkout the right branches and run the test suite with the appropriate versions.

This could be implemented by leaving tags in the PR description (the first message in the PR) and using the excellent github golang library from google: https://github.com/google/go-github.

**Comment by sameo** [comment]
Another alternative would be to implement those tests as a library, under a `runtime/pkg/tests` package.

The `tests` repo would only consume this package in order to run them. This would limit the possible deadlocks you're describing significantly.

**Comment by devimc** [comment]
I'm confused, if the objective was to avoid deadlocks then why all components were separated?
A deadlock could even occur between components. For example a PR to change the cc-proxy protocol will require another PR in cc-shim to adapt this this protocol but this last PR depends of PR that has not been merged! (deadlock!)(the argument of 'deadlock' is something contradictory). If we really want to avoid deadlocks then let's keep all components together again!

**Comment by sameo** [comment]
@devimc You're right that having separate components could trigger those deadlocks The reason for separating them is that they're logically separated components and are linked together through a protocol. If we can't design a protocol flexible enough to evolve in a backward compatible way, we're doing a bad job at it, and putting all those components in one repo is only going to hide that.

For tests we should be more flexible and we don't want to force ourselves into writing tests that would be statically defined and could never evolve, even in a non backward compatible way. So we should find a way to do so while at the same time being able to run those tests for several components.

**Comment by devimc** [comment]
My points:
1. functional test must be part of each components, this repo should not contain any functional test since them depends of each component.
2. this is not a temporal repo, we can use it to write integration test (all components working as as whole).
3. the idea of having a test library is not too bad, this library could include a 'Table Driven Tests' to see what commands are supported and what is the expected input and output.
4. Integration tests with docker and CRI-O will never change since it is code that we can't touch therefore if you PR fails with a ```docker run``` you did something wrong and you PR shouldn't be merged

**Comment by sameo** [comment]
@devimc Agreed with points 2,3 and 4. On point 1, functional tests can include more than one component in which case we should have them in our test library.

**Comment by chavafg** [comment]
Agree with the last comments from @devimc and @sameo. The functional tests that involve more than one component, should live here along with the integration tests for docker, cri-o... 
And we shouldn't have deadlocks since this will be black box testing. The way the code is modified internally, should not affect the result of the test.

**Comment by sboeuf** [comment]
Dumb question: what is the difference between functional and integration tests ? I guess functional stands for unit testing while integration is about testing the whole chain ?
I ask because functional or integration, any failing test should prevent a PR to be merged, right ?

**Comment by devimc** [comment]
@sboeuf according with wikipedia (and I agree with that) 
```
Functional testing usually describes what the system does.
Functional testing tests a slice of functionality of the whole system
```
system="Clear Containers"

**Comment by dlespiau** [comment]
we're reusing the definitions established in the 1.x/2.0 time frame and they may not have aged well. We have used them for:

- functional testing: test the behaviour of the runtime binary
- integration testing: test the whole system when CC is integrated with Docker/CRI-O/k8s

**Comment by sameo** [comment]
afaict our functional tests are not standalone ones.

**Comment by chavafg** [comment]
@sboeuf well... 
If we go to the books, unit testing is intended to verify that the individual units of the source code work correctly, for example a function. On the other hand, Functional testing has no idea of how a module is coded, functional testing is designed with an specification and it is a type of black box testing (we treat a module as a black box).

In CC3.0 we may have 2 types of functional testing. 1. Functional testing for a component, and 2. functional testing that involves more than one component. The second one, we can also call it integration testing, as we are testing the integration of all the components.  In addition we will have the integrations tests for docker, crio, k8s.

This is where we need to agree, so we do not get confused.

**Pull request closed without merging**